common/src/Microsoft.Azure.IIoT.AspNetCore/src/Auth/Clients/BehalfOfTokenProvider.cs

```diff
@@ -83,13 +83,25 @@ public partial class BehalfOfTokenProvider : ITokenProvider {
             var cache = _store.GetCache($"OID:{user.GetObjectId()}");
             var ctx = CreateAuthenticationContext(_config.InstanceUrl,
                 _config.TenantId, cache);
-            try {
-                var result = await ctx.AcquireTokenAsync(resource,
-                    new ClientCredential(_config.AppId, _config.AppSecret),
-                    new UserAssertion(token, kGrantType, name));
+
+            try {   
+                var result = await ctx.AcquireTokenSilentAsync(resource, _config.AppId);
                 return result.ToTokenResult();
             }
             catch (AdalException ex) {
+                if (ex.ErrorCode == AdalError.FailedToAcquireTokenSilently) {
+                    try {           
+                        var result = await ctx.AcquireTokenAsync(resource,
+                        new ClientCredential(_config.AppId, _config.AppSecret),
+                        new UserAssertion(token, kGrantType, name));
+                        return result.ToTokenResult();
+                    }
+                    catch (AdalException ex1) {
+                        _handler.Handle(_ctx.HttpContext, new AuthenticationException(
+                        $"Failed to authenticate on behalf of {name}", ex1));
+                        return null;
+                    }
+                }
                 _handler.Handle(_ctx.HttpContext, new AuthenticationException(
```

Comment:
This handler is what we should use.   It was intended to do the non-silent auth.   However, what I would suggest is to change it to remove the _handler entirely and have instead a bool in the constructor to enable this new code.  Set the bool to false by default (bool acquireTokenIfSilentFails = false) to be backcompat, and set it explicitly to true when adding the provider in the Autofac container in Startup.cs.

Reply:
changed